tests/test_io.py

```diff
@@ -8,86 +8,305 @@
+def tree(tmpdir, request):
+    r"""Create file tree."""
+
+    files = request.param
+    paths = []
+
+    for path in files:
+        path = path.replace('/', os.path.sep)
+        if path.endswith(os.path.sep):
+            path = audeer.path(tmpdir, path)
+            path = audeer.mkdir(path)
+            path = path + os.path.sep
+            paths.append(path)
+        else:
+            path = audeer.path(tmpdir, path)
+            audeer.mkdir(os.path.dirname(path))
+            path = audeer.touch(path)
+            paths.append(path)
+
+    yield paths
+
+
+@pytest.mark.parametrize(
+    'tree, root, files, archive_create, archive_extract, destination, '
+    'expected',
+    [
+        (  # empty
+            [],
+            '.',
+            [],
+            'archive.zip',
+            'archive.zip',
+            '.',
+            [],
+        ),
+        (
+            [],
+            '.',
+            None,
+            'archive.zip',
+            'archive.zip',
+            '.',
+            [],
+        ),
+        (  # single file
+            ['file.txt'],
+            '.',
+            'file.txt',
+            'archive.zip',
+            'archive.zip',
+            '.',
+            ['file.txt'],
+        ),
+        (  # archive folder does not exist
+            ['file.txt'],
+            '.',
+            'file.txt',
+            'sub/archive.zip',
+            'sub/archive.zip',
+            '.',
+            ['file.txt'],
+        ),
+        (  # file in sub folder
+            ['file.txt', 'sub/a/b/file.txt'],
+            '.',
+            ['sub/a/b/file.txt', 'file.txt'],
+            'archive.zip',
+            'archive.zip',
+            '.',
+            ['sub/a/b/file.txt', 'file.txt'],
+        ),
+        (
+            ['file.txt', 'sub/a/b/file.txt'],
+            '.',
+            ['sub/a/b/file.txt'],
+            'archive.zip',
+            'archive.zip',
+            '.',
+            ['sub/a/b/file.txt'],
+        ),
+        (
+            ['file.txt', 'sub/a/b/file.txt'],
+            '.',
+            ['file.txt'],
+            'archive.zip',
+            'archive.zip',
+            '.',
+            ['file.txt'],
+        ),
+        (  # include all files
+            ['file.txt', 'sub/a/b/file.txt', '.hidden'],
+            '.',
+            None,
+            'archive.zip',
+            'archive.zip',
+            '.',
+            ['.hidden', 'file.txt', 'sub/a/b/file.txt'],
+        ),
+        (  # tar.gz
+            ['file.txt', 'sub/a/b/file.txt'],
+            '.',
+            ['sub/a/b/file.txt', 'file.txt'],
+            'archive.tar.gz',
+            'archive.tar.gz',
+            '.',
+            ['sub/a/b/file.txt', 'file.txt'],
+        ),
+        (  # root is sub folder
+            ['sub/file.txt'],
+            './sub',
+            ['file.txt'],
+            'archive.zip',
+            'archive.zip',
+            '.',
+            ['file.txt'],
+        ),
```

Comment:
I would repeat this test, but don't provide `files` to see if it is able to figure out the correct names without including `root`:

```suggestion
        (  # root is sub folder
            ['sub/file.txt'],
            './sub',
            ['file.txt'],
            'archive.zip',
            'archive.zip',
            '.',
            ['file.txt'],
        ),
        (
            ['sub/file.txt'],
            './sub',
            None,
            'archive.zip',
            'archive.zip',
            '.',
            ['file.txt'],
        ),
```